### PR TITLE
[2.8] Fix check for exact version in Helm index

### PR DESCRIPTION
**Pull request by weyfonk**
When installing a system chart by version, Rancher now errors if the version of the chart found in the index is not the requested one. In such a case, the chart is not installed. This previously happened only when the version to be installed would overwrite whatever other version was already installed.

Broadening the scope of this check fixes a bug whereby a system chart, eg. Fleet, would be installed in its latest available version if an invalid version number was specified in environment variables (eg. `CATTLE_FLEET_VERSION`).

## Issue:

* [fleet#1590](https://github.com/rancher/fleet/issues/1590) and [fleet#1948](https://github.com/rancher/fleet/issues/1948)
* https://github.com/rancher/fleet/issues/1948
 
## Problem

Rancher installs a version of a Helm chart which differs from the version specified in the corresponding environment variable. The installed version is not necessarily the latest available one either, which causes confusion.
 
## Solution

Expand the check for availability of the desired version to any case of a system chart installation.
 
### Manual Testing

Ran Rancher in VS Code, setting env var `CATTLE_FLEET_VERSION` to `102.2.0+up0.8.4` (nonexistent version in `rancher/charts`' `dev-v2.8` branch at the time of testing)

#### Before the change

Could see Rancher install Fleet 0.8.0 (screenshot by @sbulage , before I reproduced the issue on my machine):
image

#### After the change

Could see Fleet not being installed by Rancher, which would emit error messages such as:
```
2023/11/06 12:15:03 [ERROR] Failed to install system chart fleet-crd: specified exact version 102.2.0+up0.8.4 doesn't exist in the index
2023/11/06 12:15:03 [ERROR] Failed to install system chart fleet: specified exact version 102.2.0+up0.8.4 doesn't exist in the index
```

## QA Testing Considerations

I have also manually tested installing an existing version (happy path), which seemed to still work as expected.
 
### Regressions Considerations
N/A

### CODE REVIEW

File: pkg/catalogv2/system/system.go [hunk 1]
```diff
@@ -266,8 +266,12 @@ func (m *Manager) install(namespace, name, minVersion, exactVersion string, valu
 	if err != nil {
 		return err
 	}
-	v := ">=0-a" // latest - this is special syntax to match everything including pre-releases build
+
+	const latestVersionMatcher = ">=0-a" // latest - special syntax to match everything including pre-release builds
```
Comment (KevinJoiner) [inline pkg/catalogv2/system/system.go:270]: This is good, but I noticed that the comment for this function was not updated to reflect the behavior change.

> // install tries to install a new version of a chart. If the exact version is provided, it will try to install it
// otherwise it will try to install the latest version available. 

Since it was already wrong, you do not have to fix it in this PR. But if you don't mind, we could squeeze it in.

**Review by KevinJoiner — DISMISSED** [review]
LGTM. I had one suggestion about updating the comment, but this is good, with or without that change.

File: pkg/catalogv2/system/system.go (hunk 1 above)
Comment (weyfonk) [reply pkg/catalogv2/system/system.go:270]: Better late than never, and better now than even later :) Thanks for raising this, I've just updated the comment. Feel free to let me know if you had something else in mind.

File: pkg/catalogv2/system/system.go [hunk 2]
```diff
@@ -285,8 +289,8 @@ func (m *Manager) install(namespace, name, minVersion, exactVersion string, valu
 		return fmt.Errorf(err.Error())
 	}
 	// Because of the behavior of `index.Get`, we need this check.
-	if isExact && chart.Version != v {
-		return fmt.Errorf("specified exact version %s doesn't exist in the index", exactVersion)
+	if v != latestVersionMatcher && chart.Version != v {
+		return fmt.Errorf("specified exact version %s doesn't exist in the index", v)
```
[the remark this answers is not in the document]
Comment (weyfonk) [reply pkg/catalogv2/system/system.go:293]: > the chart.Version could be above that version since it's just a min
You are right, but the idea is that despite the naming which survives through these variables, we no longer implement actual _min version_ behaviour: both `minVersion` and `exactVersion` are treated as exact versions, with an error being thrown (from this PR on) whenever a release cannot be sourced for that version.
The current difference between `minVersion` and `exactVersion` is that `exactVersion`, when set on its own, leads to Rancher overwriting any already installed version, even if more recent, with that specific exact version.

In practice, this [does not happen](https://github.com/rancher/rancher/blob/release/v2.8/pkg/controllers/dashboard/fleetcharts/controller.go#L91) as Fleet is the only component relying on min version behaviour: only `minVersion` is specified, yet room is left for maintainers of other charts to implement different behaviour as they see fit.

TL;DR: leaving `isExact` here would mean that this condition would not apply to calls triggered by Fleet chart installs, which are the only use case for min version at the moment. Therefore, Rancher would still install a Fleet version different from that specified in `CATTLE_FLEET_VERSION` if that configured version is not found.

File: pkg/catalogv2/system/system.go [hunk 3]
```diff
@@ -257,17 +257,21 @@ func (m *Manager) Remove(namespace, name string) {
 	}
 }
 
-// install tries to install a new version of a chart. If the exact version is provided, it will try to install it
-// otherwise it will try to install the latest version available. If a release with the version to be installed is already installed,
-// or it's pending install, upgrade or rollback this does nothing.
+// install tries to install a new version of a chart. If the exact version is provided, it will try to install it;
+// otherwise it will return an error without altering existing state. If a release with the version to be installed is
+// already installed, or is pending install, upgrade or rollback, this does nothing.
```
Comment (KevinJoiner) [inline pkg/catalogv2/system/system.go:262]: This update makes it sound like the function always returns an error if the exact version is not provided.
```suggestion
// install tries to install a new version of a chart. If the exact version is provided, it will try to install it.
// If minVersion is provided, it will try to install it only if the current version is earlier than minVersion.
// otherwise, it will try to install the latest version available. if a release with the version to be installed is
// already installed, or is pending install, upgrade or rollback, this does nothing.
```

File: pkg/catalogv2/system/system.go (hunk 2 above)
Comment (KevinJoiner) [inline pkg/catalogv2/system/system.go:293]:
```suggestion
		return fmt.Errorf("specified version %s doesn't exist in the index", v)
```

**Review by KevinJoiner — CHANGES_REQUESTED** [review]

File: pkg/catalogv2/system/system.go (hunk 3 above)
Comment (weyfonk) [reply pkg/catalogv2/system/system.go]: Thanks, included this suggestion with a few tweaks as discussed offline.

File: pkg/catalogv2/system/system.go (hunk 2 above)
[the remark this answers is not in the document]
Comment (cbron) [reply pkg/catalogv2/system/system.go]: Ok, I'm fine with this logic, but the comment around minVersion says: 
```
		// not enforcing exact version install, to keep any possibly newer, already installed version.
		// This prevents automated downgrades of updates done manually through the web UI, eg. for Fleet.
```
Which is either incorrect or confusing. Overall I'd like to see the above context reflected in the comments in this method.

[the remark this answers is not in the document]
Comment (KevinJoiner) [reply pkg/catalogv2/system/system.go]: > Which is either incorrect or confusing.
The comment is still true, with or without this change. I think you are confused about what `chart.Version` represents. This change does not enforce that the minVersion is the version installed. It only enforces that the minVersion exists in the repository.

> Overall, I'd like to see the above context reflected in the comments in this method.
Do the changes to the function comment satisfy this request? If not, can you be more specific on what comments you are looking for.

**Review by KevinJoiner — DISMISSED** [review]

File: pkg/catalogv2/system/system.go (hunk 2 above)
[the remark this answers is not in the document]
Comment (weyfonk) [reply pkg/catalogv2/system/system.go]: Would something like do?
```
@@ -281,7 +281,7 @@ func (m *Manager) install(namespace, name, minVersion, exactVersion string, valu
                isExact = true
        } else if minVersion != "" {
                v = minVersion
-               // not enforcing exact version install, to keep any possibly newer, already installed version.
+               // not setting isExact, to keep any possibly newer, already installed version.
                // This prevents automated downgrades of updates done manually through the web UI, eg. for Fleet.
        }
 
@@ -292,7 +292,10 @@ func (m *Manager) install(namespace, name, minVersion, exactVersion string, valu
                // The helm library is using github.com/pkg/errors which is deprecated
                return fmt.Errorf(err.Error())
        }
+
        // Because of the behavior of `index.Get`, we need this check.
+       // Regardless of whether `minVersion` or `exactVersion` applies, we are looking for an exact version match in
+       // the charts index.
        if v != latestVersionMatcher && chart.Version != v {
                return fmt.Errorf("specified version %s doesn't exist in the index", v)
        }
```

[the remark this answers is not in the document]
Comment (weyfonk) [reply pkg/catalogv2/system/system.go]: To reduce confusion, I would also happily rename `isExact` to `overwriteNewer` or similar if that helps.

[the remark this answers is not in the document]
Comment (cbron) [reply pkg/catalogv2/system/system.go]: That comment at top of method helps. I talked with Kevin and I mostly understand it now. I think a bigger refactor would be good, but that goes beyond scope of this issue.

**Comment by weyfonk** [comment]
I do not have sufficient permissions to merge this myself; @KevinJoiner or @cbron could you please do it? Thanks :smiley:

**Comment by cbron** [comment]
You should have permissions to merge, it's just failing CI. Also you should squash all your commits to a single one.

**Comment by diogoasouza** [comment]
Would it be possible to add integration tests for this? You could either add them [in this suite](https://github.com/rancher/rancher/blob/release/v2.8/tests/v2/integration/catalogv2/rancher_managed_charts_test.go) or in [this one](https://github.com/rancher/rancher/blob/release/v2.8/tests/v2/integration/catalogv2/system_charts_version_test.go). The latter one is disabled due to some problems in the rancher-webhook test, but you could enable the suit and disable only that specific test.

**Comment by weyfonk** [comment]
> Would it be possible to add integration tests for this? You could either add them [in this suite](https://github.com/rancher/rancher/blob/release/v2.8/tests/v2/integration/catalogv2/rancher_managed_charts_test.go) or in [this one](https://github.com/rancher/rancher/blob/release/v2.8/tests/v2/integration/catalogv2/system_charts_version_test.go). The latter one is disabled due to some problems in the rancher-webhook test, but you could enable the suit and disable only that specific test.

Integration test cases added to the system charts suite, where the webhook test case is now commented out.
The part I am unsure about is setting a Fleet version as a prerequisite to running the tests, because:
A. Downgrading Fleet is not supported, hence integration tests must set a version lower than the latest one, when setting up the test suite, to be able to test installation of higher versions
B. Prior to execution of integration tests, might something else set the Fleet version to the latest one (eg. via `build.yaml`), which would then prevent these Fleet install tests from succeeding due to A.?

I have run these integration tests on their own with a VS Code-powered Rancher instance, but would be interested to know how/if those test suites run in CI.

**Review by KevinJoiner — APPROVED** [review]

**Comment by diogoasouza** [comment]
> > Would it be possible to add integration tests for this? You could either add them [in this suite](https://github.com/rancher/rancher/blob/release/v2.8/tests/v2/integration/catalogv2/rancher_managed_charts_test.go) or in [this one](https://github.com/rancher/rancher/blob/release/v2.8/tests/v2/integration/catalogv2/system_charts_version_test.go). The latter one is disabled due to some problems in the rancher-webhook test, but you could enable the suit and disable only that specific test.
> 
> Integration test cases added to the system charts suite, where the webhook test case is now commented out. The part I am unsure about is setting a Fleet version as a prerequisite to running the tests, because: A. Downgrading Fleet is not supported, hence integration tests must set a version lower than the latest one, when setting up the test suite, to be able to test installation of higher versions B. Prior to execution of integration tests, might something else set the Fleet version to the latest one (eg. via `build.yaml`), which would then prevent these Fleet install tests from succeeding due to A.?
> 
> ~I have run these integration tests on their own with a VS Code-powered Rancher instance, but would be interested to know how/if those test suites run in CI.~ Edit: they are run via Drone, and currently failing, most likely because the Fleet version setting points to `103.1.0+up0.9.0`, ie the latest version, preventing any other version from being installed.
> 
> Btw, this required the following hack (uncommitted, see below) to work with VS Code, because:
> 
> * environment variables set in `launch.json` had precedence over settings updates made via the test framework
> * my understanding is that running a Rancher instance from VS Code does not involve a Docker container for Rancher itself, therefore variables from `build.yaml` are not used as build arguments. In Fleet's case, this means that installation defaults to [a latest version matcher](https://github.com/rancher/rancher/blob/release/v2.9/pkg/catalogv2/system/system.go#L269).
> 
> ```
> diff --git a/pkg/controllers/dashboard/fleetcharts/controller.go b/pkg/controllers/dashboard/fleetcharts/controller.go
> index 8b3ba8588..b37e41fbb 100644
> --- a/pkg/controllers/dashboard/fleetcharts/controller.go
> +++ b/pkg/controllers/dashboard/fleetcharts/controller.go
> @@ -6,6 +6,7 @@
>         "sync"
>  
>         v3 "github.com/rancher/rancher/pkg/apis/management.cattle.io/v3"
> +       "github.com/rancher/rancher/pkg/buildconfig"
>         "github.com/rancher/rancher/pkg/controllers/dashboard/chart"
>         "github.com/rancher/rancher/pkg/features"
>         fleetconst "github.com/rancher/rancher/pkg/fleet"
> @@ -88,6 +89,11 @@ func (h *handler) onSetting(key string, setting *v3.Setting) (*v3.Setting, error
>                 fleetVersion = fleetMinVersion
>         }
>  
> +       if fleetVersion == "" {
> +               // Set the Fleet version from build defaults in case no environment variables are set.
> +               fleetVersion = buildconfig.FleetVersion
> +       }
> +
>         err := h.manager.Ensure(
>                 fleetCRDChart.ReleaseNamespace,
>                 fleetCRDChart.ChartName,
> ```
> 
> Happy to discuss!

If I understood it right, the issue is that the Drone pipeline sets the fleet version to the latest one in the rancher-charts repo and because of that you can't perform an upgrade, right? 
I had a similar problem (had to test an upgrade to a _broken_ version) when creating another integration test, and one way I found able to do it was to change the repository the rancher-charts CRD pointed to and in that repository, I had different chart versions to be able to test it. If you want to take a look at it, it's [here](https://github.com/rancher/rancher/blob/5d4be30e9286f1e67be0805ccb622b1b3e2bce23/tests/v2/integration/catalogv2/rancher_managed_charts_test.go#L193), maybe with that you won't need that hack

**Comment by diogoasouza** [comment]
> > If I understood it right, the issue is that the Drone pipeline sets the fleet version to the latest one in the rancher-charts repo and because of that you can't perform an upgrade, right? I had a similar problem (had to test an upgrade to a _broken_ version) when creating another integration test, and one way I found able to do it was to change the repository the rancher-charts CRD pointed to and in that repository, I had different chart versions to be able to test it. If you want to take a look at it, it's [here](https://github.com/rancher/rancher/blob/5d4be30e9286f1e67be0805ccb622b1b3e2bce23/tests/v2/integration/catalogv2/rancher_managed_charts_test.go#L193), maybe with that you won't need that hack
> 
> Correct, that's the issue. Thanks for the pointer. In this case, we realised that this could be covered via unit tests, which needed to be updated anyway to reflect Helm's version matching behaviour (see commit [c6a67a7](https://github.com/rancher/rancher/commit/c6a67a751443bce4c54fd8c6dfcc613e8580f357)). Therefore, new integration test cases should not be needed. Thanks for raising the need for proper test coverage anyway, which triggered this effort.

Got it, in that case, the integration tests are indeed not needed.

**Review by diogoasouza — APPROVED** [review]

**Review by manno — APPROVED** [review]
